sql/rls: apply SELECT policies for UPDATE only if columns are referenced

spilchen opened the pull request:
Previously, SELECT policies were applied for the new updated row on every UPDATE operation, regardless of whether any columns were referenced in the SET or WHERE clauses. This behavior was inconsistent with postgres, where SELECT policies should only be applied if an UPDATE references a column in either the SET or WHERE expression.

Note: This change does not yet handle the RETURNING clause, which will be addressed separately.

Fixes #144943

Epic: CRDB-11724
Release note (bug fix): SELECT policies during UPDATE operations are now only applied when referenced columns appear in the SET or WHERE clauses, matching postgres' behavior. This improves compatibility.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145344)
<!-- Reviewable:end -->

mgartner reviewed (commented):
As I was reviewing, I noticed that we need to so something similar to the SELECT policy filters applied during an UPDATE. I created #145894 to track this—sorry if there is already a duplicate issue somewhere.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-:-OPBJ4tz0kra0y5fNt6a:b-in7vgl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 4397 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPB9Nkr3L3h8X85C2fR:-OPB9Nkr3L3h8X85C2fS:buaje2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/885e34eae9e91048e942046cd99a04321d38025d/pkg/sql/logictest/testdata/logic_test/row_level_security#L4397)):*
> ```
> 
> statement ok
> UPDATE t SET a = 0 WHERE true;
> ```

Can you add an `UPDATE t SET a = -1 WHERE true` immediately after this? It should succeed based on my tests with Postgres, though I think right now it will update zero rows due to #145894, so you can leave a TODO for now.

```sql
DROP TABLE IF EXISTS t;
DROP ROLE IF EXISTS alice;

CREATE ROLE alice;

CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT
);

INSERT INTO t VALUES (1, 0);

GRANT SELECT, INSERT, UPDATE, DELETE ON t TO alice;

ALTER TABLE t ENABLE ROW LEVEL SECURITY;

CREATE POLICY select_policy_alice
ON t
FOR SELECT
TO alice
USING (a > 0);

CREATE POLICY update_policy_alice
ON t
FOR UPDATE
TO alice
USING (true);

SET ROLE alice;

-- a appears in the SET clause but it denotes a new value to write, and it does
-- not read the existing value.
UPDATE t SET a = -1 WHERE true;
-- In Postgres, this update succeeds:
-- UPDATE 1
--
-- In CRDB with #144943, the update affects no rows.
-- UPDATE 0

SET ROLE demo;
-- SET ROLE marcus;

SELECT * FROM t;
-- Postgres:
--  k | a
-- ---+----
--  1 | -1
-- (1 row)
--
-- CRDB:
--  k | a
-- ---+----
--  1 | 0
-- (1 row)
```
___
*[`pkg/sql/opt/optbuilder/update.go` line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPBBoFY9AJsica5-CE5:-OPBBoFY9AJsica5-CE6:b-2cryvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/885e34eae9e91048e942046cd99a04321d38025d/pkg/sql/opt/optbuilder/update.go#L362)):*
> ```Go
> 	mb.addCheckConstraintCols(true, /* isUpdate */
> 		policyScopeCmd,
> 		returning != nil || (colRefs != nil && colRefs.Len() > 0) /* includeSelectPolicies */)
> ```


nit: the expression is complex enough to be its own variable, e.g., `includeSelectPolicies := returning != nil || ...`

----

I think checking that the colRefs set is non-empty isn't enough, e.g., if we have an implicit JOIN with `UPDATE ... FROM` we might reference columns from the table not being updated, in which case the updating table's select policies should not apply.

Here's an example with SET:

```sql
DROP TABLE IF EXISTS t;
DROP TABLE IF EXISTS t2;
DROP ROLE IF EXISTS alice;

CREATE ROLE alice;

CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT
);

INSERT INTO t VALUES (1, 10);

GRANT SELECT, INSERT, UPDATE, DELETE ON t TO alice;

ALTER TABLE t ENABLE ROW LEVEL SECURITY;

CREATE POLICY select_policy_alice
ON t
FOR SELECT
TO alice
USING (a > 0);

CREATE POLICY update_policy_alice
ON t
FOR UPDATE
TO alice
USING (true);

CREATE TABLE t2 (k INT PRIMARY KEY, a INT);

INSERT INTO t2 VALUES (1, -1);

GRANT SELECT, INSERT, UPDATE, DELETE ON t2 TO alice;

SET ROLE alice;

UPDATE t SET a = t2.a FROM t2 WHERE true;
-- In Postgres, this update succeeds:
-- UPDATE 1
--
-- In CRDB with #144943, the update errors.
-- ERROR: new row violates row-level security policy for table "t"
```

And here's an example with WHERE:

```sql
DROP TABLE IF EXISTS t;
DROP TABLE IF EXISTS t2;
DROP ROLE IF EXISTS alice;

CREATE ROLE alice;

CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT
);

INSERT INTO t VALUES (1, 10);

GRANT SELECT, INSERT, UPDATE, DELETE ON t TO alice;

ALTER TABLE t ENABLE ROW LEVEL SECURITY;

CREATE POLICY select_policy_alice
ON t
FOR SELECT
TO alice
USING (a > 0);

CREATE POLICY update_policy_alice
ON t
FOR UPDATE
TO alice
USING (true);

CREATE TABLE t2 (k INT PRIMARY KEY, a INT);

INSERT INTO t2 VALUES (1, -1);

GRANT SELECT, INSERT, UPDATE, DELETE ON t2 TO alice;

SET ROLE alice;

UPDATE t SET a = -2 FROM t2 WHERE t2.k = 1;
-- In Postgres, this update succeeds:
-- UPDATE 1
--
-- In CRDB with #144943, the update errors.
-- ERROR: new row violates row-level security policy for table "t"
```


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/optbuilder/update.go` line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPBBoFY9AJsica5-CE5:-OPBNFvr3Rc9T478rY6O:b-3u536q) ([raw file](https://github.com/cockroachdb/cockroach/blob/885e34eae9e91048e942046cd99a04321d38025d/pkg/sql/opt/optbuilder/update.go#L362)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: the expression is complex enough to be its own variable, e.g., `includeSelectPolicies := returning != nil || ...`

----

I think checking that the colRefs set is non-empty isn't enough, e.g., if we have an implicit JOIN with `UPDATE ... FROM` we might reference columns from the table not being updated, in which case the updating table's select policies should not apply.

Here's an example with SET:

```sql
DROP TABLE IF EXISTS t;
DROP TABLE IF EXISTS t2;
DROP ROLE IF EXISTS alice;

CREATE ROLE alice;

CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT
);

INSERT INTO t VALUES (1, 10);

GRANT SELECT, INSERT, UPDATE, DELETE ON t TO alice;

ALTER TABLE t ENABLE ROW LEVEL SECURITY;

CREATE POLICY select_policy_alice
ON t
FOR SELECT
TO alice
USING (a > 0);

CREATE POLICY update_policy_alice
ON t
FOR UPDATE
TO alice
USING (true);

CREATE TABLE t2 (k INT PRIMARY KEY, a INT);

INSERT INTO t2 VALUES (1, -1);

GRANT SELECT, INSERT, UPDATE, DELETE ON t2 TO alice;

SET ROLE alice;

UPDATE t SET a = t2.a FROM t2 WHERE true;
-- In Postgres, this update succeeds:
-- UPDATE 1
--
-- In CRDB with #144943, the update errors.
-- ERROR: new row violates row-level security policy for table "t"
```

And here's an example with WHERE:

```sql
DROP TABLE IF EXISTS t;
DROP TABLE IF EXISTS t2;
DROP ROLE IF EXISTS alice;

CREATE ROLE alice;

CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT
);

INSERT INTO t VALUES (1, 10);

GRANT SELECT, INSERT, UPDATE, DELETE ON t TO alice;

ALTER TABLE t ENABLE ROW LEVEL SECURITY;

CREATE POLICY select_policy_alice
ON t
FOR SELECT
TO alice
USING (a > 0);

CREATE POLICY update_policy_alice
ON t
FOR UPDATE
TO alice
USING (true);

CREATE TABLE t2 (k INT PRIMARY KEY, a INT);

INSERT INTO t2 VALUES (1, -1);

GRANT SELECT, INSERT, UPDATE, DELETE ON t2 TO alice;

SET ROLE alice;

UPDATE t SET a = -2 FROM t2 WHERE t2.k = 1;
-- In Postgres, this update succeeds:
-- UPDATE 1
--
-- In CRDB with #144943, the update errors.
-- ERROR: new row violates row-level security policy for table "t"
```
</blockquote></details>

I think we can fix this with something like:

```go
	includeSelectPolicies := returning != nil
	if !includeSelectPolicies {
		for _, colID := range mb.fetchColIDs {
			if colID != 0 && colRefs.Contains(colID) {
				includeSelectPolicies = true
				break
			}
		}
	}
```


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/optbuilder/update.go` line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPBBoFY9AJsica5-CE5:-OPBODvR1U_b1PPvdpEJ:b9hug2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/885e34eae9e91048e942046cd99a04321d38025d/pkg/sql/opt/optbuilder/update.go#L362)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think we can fix this with something like:

```go
	includeSelectPolicies := returning != nil
	if !includeSelectPolicies {
		for _, colID := range mb.fetchColIDs {
			if colID != 0 && colRefs.Contains(colID) {
				includeSelectPolicies = true
				break
			}
		}
	}
```
</blockquote></details>

Hmm, we should make sure that `mb.fetchColIDs` contains all of the original table's columns during the build phase, even if they are not updated or referenced, otherwise my suggestion about won't work.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/optbuilder/update.go` line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPBBoFY9AJsica5-CE5:-OPBU4Lj7zJbwfkvyytr:be45zjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/885e34eae9e91048e942046cd99a04321d38025d/pkg/sql/opt/optbuilder/update.go#L362)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Hmm, we should make sure that `mb.fetchColIDs` contains all of the original table's columns during the build phase, even if they are not updated or referenced, otherwise my suggestion about won't work.
</blockquote></details>

I did some quick testing and I think using `mb.fetchColIDs` here should be fine.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145344)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/optbuilder/update.go` line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPBBoFY9AJsica5-CE5:-OPFntSX5S28KD9BmfE_:b-k1cv0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/885e34eae9e91048e942046cd99a04321d38025d/pkg/sql/opt/optbuilder/update.go#L362)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I did some quick testing and I think using `mb.fetchColIDs` here should be fine.
</blockquote></details>

Another good find! I applied the suggested change and added test coverage for this scenario in both the logic and optbuilder tests.
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 4397 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPB9Nkr3L3h8X85C2fR:-OPFgPNA6cXOGA3fofaM:b-xp263r) ([raw file](https://github.com/cockroachdb/cockroach/blob/885e34eae9e91048e942046cd99a04321d38025d/pkg/sql/logictest/testdata/logic_test/row_level_security#L4397)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add an `UPDATE t SET a = -1 WHERE true` immediately after this? It should succeed based on my tests with Postgres, though I think right now it will update zero rows due to #145894, so you can leave a TODO for now.

```sql
DROP TABLE IF EXISTS t;
DROP ROLE IF EXISTS alice;

CREATE ROLE alice;

CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT
);

INSERT INTO t VALUES (1, 0);

GRANT SELECT, INSERT, UPDATE, DELETE ON t TO alice;

ALTER TABLE t ENABLE ROW LEVEL SECURITY;

CREATE POLICY select_policy_alice
ON t
FOR SELECT
TO alice
USING (a > 0);

CREATE POLICY update_policy_alice
ON t
FOR UPDATE
TO alice
USING (true);

SET ROLE alice;

-- a appears in the SET clause but it denotes a new value to write, and it does
-- not read the existing value.
UPDATE t SET a = -1 WHERE true;
-- In Postgres, this update succeeds:
-- UPDATE 1
--
-- In CRDB with #144943, the update affects no rows.
-- UPDATE 0

SET ROLE demo;
-- SET ROLE marcus;

SELECT * FROM t;
-- Postgres:
--  k | a
-- ---+----
--  1 | -1
-- (1 row)
--
-- CRDB:
--  k | a
-- ---+----
--  1 | 0
-- (1 row)
```
</blockquote></details>

Good catch! I left a comment to address this later.


<!-- Sent from Reviewable.io -->

spilchen commented:
@marcus PTAL

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nicely done!

> Reviewed 7 of 10 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-:-OPVfbeE0OiHf0Eb7Tuh:b-msncp4)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1037 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPVejED3V2b3rpbwvWc:-OPVejED3V2b3rpbwvWd:b-2zhxdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/70017f5dce23bb1dae1b73bfd8f26ccdf3a28f42/pkg/sql/opt/optbuilder/mutation_builder.go#L1037)):*
> ```Go
> 			selectPolicyExpr,
> 			mb.genPolicyWithCheckExpr(tabMeta, cat.PolicyScopeUpdate, referencedCols),
> 		)
> ```

nit: `true AND <update_policy>` will get normalized to `<update_policy>`, but you can save some cycles and avoid calling `ConstructAnd` if `includeSelectPolicies` is `false`, e.g.:


```go
scalar = mb.genPolicyWithCheckExpr(tabMeta, cat.PolicyScopeUpdate, referencedCols)
if includeSelectPolicies {
    scalar = mb.b.factory.ConstructAnd(
        mb.genPolicyUsingExpr(tabMeta, cat.PolicyScopeSelect, mb.outScope, referencedCols),
        scalar,
    )
}
```
___
*[`pkg/sql/opt/optbuilder/testdata/row_level_security` line 327 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPVcy2cEg_m4_pv-kx7:-OPVcy2cEg_m4_pv-kx8:b917c6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/70017f5dce23bb1dae1b73bfd8f26ccdf3a28f42/pkg/sql/opt/optbuilder/testdata/row_level_security#L327)):*
> ```
> ----
> 
> build
> ```

nit: a comment to explain these other cases would be helpful


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145344)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1037 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145344#-OPVejED3V2b3rpbwvWc:-OPWHsUCDILJqli5TvMP:b-6ah6jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/70017f5dce23bb1dae1b73bfd8f26ccdf3a28f42/pkg/sql/opt/optbuilder/mutation_builder.go#L1037)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: `true AND <update_policy>` will get normalized to `<update_policy>`, but you can save some cycles and avoid calling `ConstructAnd` if `includeSelectPolicies` is `false`, e.g.:


```go
scalar = mb.genPolicyWithCheckExpr(tabMeta, cat.PolicyScopeUpdate, referencedCols)
if includeSelectPolicies {
    scalar = mb.b.factory.ConstructAnd(
        mb.genPolicyUsingExpr(tabMeta, cat.PolicyScopeSelect, mb.outScope, referencedCols),
        scalar,
    )
}
```
</blockquote></details>

Done. I applied this to the insert case too as it was very similar.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/rls: apply SELECT policies for UPDATE only if columns are referenced

Previously, SELECT policies were applied for the new updated row on
every UPDATE operation, regardless of whether any columns were
referenced in the SET or WHERE clauses. This behavior was inconsistent
with postgres, where SELECT policies should only be applied if an UPDATE
references a column in either the SET or WHERE expression.

Note: This change does not yet handle the RETURNING clause, which will
be addressed separately.

Fixes #144943

Epic: CRDB-11724
Release note (bug fix): SELECT policies during UPDATE operations are now
only applied when referenced columns appear in the SET or WHERE clauses,
matching postgres' behavior. This improves compatibility.

spilchen commented:
TFTR!

bors r+
